### ISSUE TITLE
RestClient: Uploading Excel Files created with Apache POI

### ISSUE THREAD

**Issue by HeavyTobi** [eclipse/microprofile-rest-client#212]
Hello,

At our company we need to upload an excel file created using Apache POI. 
I didn't find any possibility to do this. 

I already tried using MultipartForm and the Workbook itself, both didn't work.

Used implementation of Rest Client: REST Easy integrated to Quarkus

Is it even possible to upload any files using the REST client? Didn't find any examples on google / stack overflow / github. If it is not possible: This is a feature request.

**Comment by rdebusscher** [comment]
You can upload a file (any binary data) using a regular REST endpoint, for example, https://javatutorial.net/java-file-upload-rest-service.

MicroProfile Rest client has nothing  to do with defining REST endpoints but calling them.  The Quarkus repository/Forum would be a better place to ask this question.

**Comment by HeavyTobi** [comment]
> 
> 
> You can upload a file (any binary data) using a regular REST endpoint, for example, https://javatutorial.net/java-file-upload-rest-service.
> 
> MicroProfile Rest client has nothing to do with defining REST endpoints but calling them. The Quarkus repository/Forum would be a better place to ask this question.

I think you misunderstood my question: I want to upload a file from the java service to another service, not upload a file to the java service

**Comment by andymc12** [comment]
It is likely possible if you were to use a custom MessageBodyWriter to convert the File object (or whatever the parameter is) to the message body output stream.  This is probably the only portable solution as most vendors have their own unique multipart implementation.

I agree that this would be a good feature to have in the MP Rest Client spec.  In general, we try to "piggy-back" off of JAX-RS (now Jakarta REST) for consistency.  JAX-RS also does not have a spec-defined multipart API, but that has been proposed for the next release.  Ideally, whatever we do in MP Rest Client would be compatible with JAX-RS.

**Comment by watermelonjam** [comment]
While it would be nice to have an implementation-neutral way of doing this as part of the spec, the current release of Quarkus, 1.7.0.Final, with a dependency on resteasy-client 4.5.6.Final, handles this just fine for me in the case of passing binary file data over to SharePoint via its REST API.  Using an `InputStream` body parameter on the MP REST interface automatically caused it to use its `InputStreamProvider` for marshalling:

```java
package com.example.sharepoint.rest;

import java.io.InputStream;

import javax.ws.rs.Consumes;
import javax.ws.rs.HeaderParam;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;

@RegisterRestClient(configKey = "sharepoint-api")
public interface MicroprofileRestClient {
	@POST
	@Path("/web/GetFolderByServerRelativeUrl('{serverRelativeUrl}')/Files/add(url='{fileName}',overwrite=true)")
	@Produces("application/json;odata=verbose")
	@Consumes(MediaType.APPLICATION_OCTET_STREAM)
	Response store(@HeaderParam("Authorization") String authorization,
			@PathParam("serverRelativeUrl") String serverRelativeUrl, @PathParam("fileName") String fileName,
			InputStream content);
}
```

**Comment by jamezp** [comment]
This should be solved in 3.1 with the upgrade to Jakarta REST 3.1. There was a `jakarta.ws.rs.core.EntityPart` introduced in Jakarta REST 3.1.

**Comment by Emily-Jiang** [comment]
similar situation to #261
